src/buildblock/FilePath.cxx

```diff
@@ -0,0 +1,384 @@
+
+    std::string new_path = my_string;
+
+    //Check if this a directory or it contains a filename, too
+    if(!is_directory())
+    {
+        if(!is_regular_file())
+        {
+            error(boost::format("FilePath: Cannot find a directory in %1%.")%my_string);
+        }
+        else
+        {
+            new_path = get_path();
+        }
+    }
+
+    // Try to accomondate multiple sub paths
+    // Find if string p has more than one levels and store them in a vector.
+    std::vector<std::string> v = split(p, separator.c_str());
+
+    // Run over the vector creating the subfolders recrusively.
+    for (unsigned int i = 0; i < v.size(); i++)
+    {
+        new_path = merge(new_path, v.at(i));
+        FilePath::append_separator(new_path);
+
+        // if current level already exists move to the next.
+        if (FilePath::exist(new_path) == true)
+        {
+            warning(boost::format("FilePath: Path %1% already exists.")%new_path);
+            continue;
+        }
+		int nError;
+#if defined(__OS_WIN__)
+		nError = _mkdir(new_path.c_str());
+#else
+		nError = mkdir(new_path.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+#endif
+
+        if (-1 == nError)
+        {
+            printf("Error creating directory!n");
+            exit(1);
+        }
+    }
+    return FilePath(new_path);
+}
+
+void
+FilePath::add_extension(const std::string &e)
+{
+    std::string::size_type pos = find_pos_of_extension();
+    if (pos == std::string::npos)
+        my_string += e;
+}
+
+void
+FilePath::replace_extension(const std::string& e)
+{
+    std::string::size_type pos = find_pos_of_extension();
+    if (pos != std::string::npos)
+    {
+        my_string.erase(pos);
+        my_string += e;
+    }
+}
+
+
+std::string
+FilePath::get_path() const
+{
+    std::size_t i = my_string.rfind(separator, my_string.length());
+    if (i != std::string::npos)
+    {
+        return(my_string.substr(0, i+1));
+    }
+
+    return(my_string);
+}
+
+std::string
+FilePath::get_filename() const
+{
+    std::size_t i = my_string.rfind(separator, my_string.length());
+
+    if (i != std::string::npos)
+    {
+        return(my_string.substr(i+1, my_string.length() - i));
+    }
+    return(my_string);
+}
+
+std::string
+FilePath::get_extension() const
+{
+    std::size_t i = find_pos_of_extension();
+    if (i != std::string::npos)
+    {
+        return(my_string.substr(i+1, my_string.length() - i));
+    }
+    return("");
+}
+
+void FilePath::checks() const
+{
+    struct stat s;
+    if( stat(my_string.c_str(),&s) == 0 )
+    {
+        if ((s.st_mode & S_IFDIR ) &&
+               ( s.st_mode & S_IFREG ) )
+        {
+            error(boost::format("FilePath: File %1% is neither a directory nor a file")%my_string);
+        }
+    }
+    else
+    {
+        error(boost::format("FilePath: Maybe %1% does not exist?")%my_string);
+    }
+}
+
+std::string FilePath::get_current_working_directory()
+{
+    char buffer[100];
```

Comment:
100 is dangerously small. At least use the const in `utilities.h` for max filename size. Alternatively,
`get_current_dir_name` seems to allocate the buffer itself (will then have to deallocate).
not sure on Windows.

Reply:
this still needs addressing. On Windows there's [GetCurrentDirectory](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364934(v=vs.85).aspx). don't bother using it though, just put it in a comment.

some comments on all this at https://stackoverflow.com/questions/2868680/what-is-a-cross-platform-way-to-get-the-current-directory. Note the one amount `getcwd` returning `NULL`.

Reply:
Yep, that why I use that pointer to guard the string